File: packages/oxlint-plugin-react-doctor/src/plugin/rules/r3f/r3f-valid-orthographic-camera.ts
import { defineRule } from "../../utils/define-rule.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";
import { getStaticObjectPropertyValue } from "../../utils/get-static-object-property-value.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import type { RuleContext } from "../../utils/rule-context.js";
import {
  ORTHOGRAPHIC_CAMERA_BOTTOM_ARGUMENT_INDEX,
  ORTHOGRAPHIC_CAMERA_FAR_ARGUMENT_INDEX,
  ORTHOGRAPHIC_CAMERA_LEFT_ARGUMENT_INDEX,
  ORTHOGRAPHIC_CAMERA_NEAR_ARGUMENT_INDEX,
  ORTHOGRAPHIC_CAMERA_RIGHT_ARGUMENT_INDEX,
  ORTHOGRAPHIC_CAMERA_TOP_ARGUMENT_INDEX,
} from "./constants.js";
import { getInvalidOrthographicCameraParameter } from "./utils/get-invalid-orthographic-camera-parameter.js";
import { getJsxAttributeExpression } from "./utils/get-jsx-attribute-expression.js";
import { getStaticNumber } from "./utils/get-static-number.js";
import { hasR3fRuntimeImport } from "./utils/has-r3f-runtime-import.js";
import { isR3fCanvas } from "./utils/is-r3f-canvas.js";
import { readStaticJsxBooleanAttribute } from "./utils/read-static-jsx-boolean-attribute.js";
import { getAuthoritativeJsxAttribute } from "../../utils/get-authoritative-jsx-attribute.js";

interface StaticR3fCameraParameter {
  readonly node: EsTreeNode;
  readonly value: number;
}

interface StaticR3fCameraParameters {
  readonly bottom: StaticR3fCameraParameter | null;
  readonly far: StaticR3fCameraParameter | null;
  readonly left: StaticR3fCameraParameter | null;
  readonly near: StaticR3fCameraParameter | null;
  readonly right: StaticR3fCameraParameter | null;
  readonly top: StaticR3fCameraParameter | null;
}

const getStaticParameter = (
  expression: EsTreeNode | null | undefined,
  context: RuleContext,
): StaticR3fCameraParameter | null => {
  if (!expression || isNodeOfType(expression, "SpreadElement")) return null;
  const value = getStaticNumber(expression, context.scopes);
  return value === null ? null : { node: expression, value };
};

const getCameraParameters = (
  node: EsTreeNodeOfType<"JSXOpeningElement">,
  context: RuleContext,
): StaticR3fCameraParameters | null => {
  if (node.attributes.some((attribute) => isNodeOfType(attribute, "JSXSpreadAttribute")))
    return null;
  if (isNodeOfType(node.name, "JSXIdentifier") && node.name.name === "orthographicCamera") {
    const argsExpression = getJsxAttributeExpression(node, "args");
    const argumentsList =
      argsExpression && isNodeOfType(argsExpression, "ArrayExpression")
        ? argsExpression.elements
        : [];
    const readParameter = (propertyName: string, argumentIndex: number) => {
      const expression = getJsxAttributeExpression(node, propertyName);
      return getStaticParameter(
        expression === undefined ? argumentsList[argumentIndex] : expression,
        context,
      );
    };
    return {
      left: readParameter("left", ORTHOGRAPHIC_CAMERA_LEFT_ARGUMENT_INDEX),
      right: readParameter("right", ORTHOGRAPHIC_CAMERA_RIGHT_ARGUMENT_INDEX),
      top: readParameter("top", ORTHOGRAPHIC_CAMERA_TOP_ARGUMENT_INDEX),
      bottom: readParameter("bottom", ORTHOGRAPHIC_CAMERA_BOTTOM_ARGUMENT_INDEX),
      near: readParameter("near", ORTHOGRAPHIC_CAMERA_NEAR_ARGUMENT_INDEX),
      far: readParameter("far", ORTHOGRAPHIC_CAMERA_FAR_ARGUMENT_INDEX),
    };
  }
  if (!isR3fCanvas(node, context)) return null;
  const orthographicAttribute = getAuthoritativeJsxAttribute(node.attributes, "orthographic");
  if (!orthographicAttribute || readStaticJsxBooleanAttribute(orthographicAttribute) !== true)
    return null;
  const cameraExpression = getJsxAttributeExpression(node, "camera");
  if (!cameraExpression || !isNodeOfType(cameraExpression, "ObjectExpression")) return null;
  return {
    left: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "left"), context),
    right: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "right"), context),
    top: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "top"), context),
    bottom: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "bottom"), context),
    near: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "near"), context),
    far: getStaticParameter(getStaticObjectPropertyValue(cameraExpression, "far"), context),
  };
};

export const r3fValidOrthographicCamera = defineRule({
  id: "r3f-valid-orthographic-camera",
  title: "Invalid R3F orthographic camera",
  category: "Correctness",
  tags: ["react-jsx-only"],
  severity: "error",
  recommendation:
    "Use distinct horizontal and vertical frustum planes and a far plane greater than near",
  create: (context: RuleContext) => {
    let importsReactThreeFiber = false;
    return {
      Program(node: EsTreeNodeOfType<"Program">) {
        importsReactThreeFiber = hasR3fRuntimeImport(node, context.scopes);
      },
      JSXOpeningElement(node: EsTreeNodeOfType<"JSXOpeningElement">) {
        if (!importsReactThreeFiber) return;
        const parameters = getCameraParameters(node, context);
        if (!parameters) return;
        const invalidParameter = getInvalidOrthographicCameraParameter(parameters);
        if (invalidParameter)
          context.report({ node: invalidParameter.node, message: invalidParameter.message });
      },
    };
  },
});

File: native/oxlint/rules/r3f-valid-orthographic-camera.rs
use oxc_ast::{ast::JSXAttributeValue, AstKind};
use oxc_diagnostics::OxcDiagnostic;
use oxc_span::GetSpan;

use crate::{
    context::{ContextHost, LintContext},
    rule::{Rule, RuleCategory, RuleInfo, RuleMeta},
};

const R3F_PUBLIC_MODULES: [&str; 5] = [
    "@react-three/fiber",
    "@react-three/fiber/legacy",
    "@react-three/fiber/native",
    "@react-three/fiber/webgpu",
    "react-three-fiber",
];
const LEFT_ARGUMENT_INDEX: usize = 0;
const RIGHT_ARGUMENT_INDEX: usize = 1;
const TOP_ARGUMENT_INDEX: usize = 2;
const BOTTOM_ARGUMENT_INDEX: usize = 3;
const NEAR_ARGUMENT_INDEX: usize = 4;
const FAR_ARGUMENT_INDEX: usize = 5;

#[derive(Debug, Default, Clone)]
pub struct R3FValidOrthographicCamera;

impl RuleMeta for R3FValidOrthographicCamera {
    const NAME: &'static str = "r3f-valid-orthographic-camera";
    const PLUGIN: &'static str = "react_doctor_native";
    const CATEGORY: RuleCategory = RuleCategory::Correctness;
    const VERSION: &'static str = "0.1.0";
    const INFO: RuleInfo = RuleInfo {
        short_description: "Validate React Three Fiber orthographic cameras.",
    };
}

impl Rule for R3FValidOrthographicCamera {
    fn should_run(&self, ctx: &ContextHost) -> bool {
        ctx.source_type().is_jsx()
    }

    fn run_once<'a>(&self, ctx: &LintContext<'a>) {
        if file_is_non_react_jsx_dialect(ctx) || !has_r3f_runtime_import(ctx) {
            return;
        }
        for node in ctx.nodes().iter() {
            let AstKind::JSXOpeningElement(opening_element) = node.kind() else {
                continue;
            };
            if opening_element.attributes.iter().any(|attribute| {
                matches!(
                    attribute,
                    oxc_ast::ast::JSXAttributeItem::SpreadAttribute(_)
                )
            }) {
                continue;
            }
            let parameters = match &opening_element.name {
                oxc_ast::ast::JSXElementName::Identifier(identifier)
                    if identifier.name == "orthographicCamera" =>
                {
                    let arguments = get_authoritative_jsx_attribute(
                        opening_element,
                        "args",
                        true,
                    )
                    .and_then(jsx_attribute_expression)
                    .and_then(|expression| {
                        let oxc_ast::ast::Expression::ArrayExpression(arguments) =
                            expression.get_inner_expression()
                        else {
                            return None;
                        };
                        Some(arguments)
                    });
                    let parameter_expression = |property_name, argument_index| {
                        get_authoritative_jsx_attribute(opening_element, property_name, true)
                            .map_or_else(
                                || {
                                    arguments
                                        .and_then(|arguments| {
                                            arguments.elements.get(argument_index)
                                        })
                                        .and_then(
                                            oxc_ast::ast::ArrayExpressionElement::as_expression,
                                        )
                                },
                                jsx_attribute_expression,
                            )
                    };
                    Some((
                        parameter_expression("left", LEFT_ARGUMENT_INDEX),
                        parameter_expression("right", RIGHT_ARGUMENT_INDEX),
                        parameter_expression("top", TOP_ARGUMENT_INDEX),
                        parameter_expression("bottom", BOTTOM_ARGUMENT_INDEX),
                        parameter_expression("near", NEAR_ARGUMENT_INDEX),
                        parameter_expression("far", FAR_ARGUMENT_INDEX),
                    ))
                }
                _ if is_r3f_canvas(opening_element, ctx) => {
                    let Some(orthographic_attribute) =
                        get_authoritative_jsx_attribute(opening_element, "orthographic", true)
                    else {
                        continue;
                    };
                    if read_static_jsx_boolean_attribute(orthographic_attribute) != Some(true) {
                        continue;
                    }
                    let Some(camera_expression) = get_authoritative_jsx_attribute(
                        opening_element,
                        "camera",
                        true,
                    )
                    .and_then(jsx_attribute_expression)
                    else {
                        continue;
                    };
                    Some((
                        get_static_object_property_value(camera_expression, "left"),
                        get_static_object_property_value(camera_expression, "right"),
                        get_static_object_property_value(camera_expression, "top"),
                        get_static_object_property_value(camera_expression, "bottom"),
                        get_static_object_property_value(camera_expression, "near"),
                        get_static_object_property_value(camera_expression, "far"),
                    ))
                }
                _ => None,
            };
            let Some((
                left_expression,
                right_expression,
                top_expression,
                bottom_expression,
                near_expression,
                far_expression,
            )) = parameters
            else {
                continue;
            };
            let left = resolve_camera_parameter(left_expression, ctx);
            let right = resolve_camera_parameter(right_expression, ctx);
            let top = resolve_camera_parameter(top_expression, ctx);
            let bottom = resolve_camera_parameter(bottom_expression, ctx);
            let near = resolve_camera_parameter(near_expression, ctx);
            let far = resolve_camera_parameter(far_expression, ctx);
            let invalid_parameter =
                if let (Some((_, left_value)), Some((right_expression, right_value))) =
                    (left, right)
                    && left_value == right_value
                {
                    Some((
                        right_expression,
                        "OrthographicCamera left and right planes must differ",
                    ))
                } else if let (Some((_, top_value)), Some((bottom_expression, bottom_value))) =
                    (top, bottom)
                    && top_value == bottom_value
                {
                    Some((
                        bottom_expression,
                        "OrthographicCamera top and bottom planes must differ",
                    ))
                } else if let (Some((_, near_value)), Some((far_expression, far_value))) =
                    (near, far)
                    && far_value <= near_value
                {
                    Some((
                        far_expression,
                        "OrthographicCamera far must be greater than near",
                    ))
                } else {
                    None
                };
            if let Some((expression, message)) = invalid_parameter {
                ctx.diagnostic(OxcDiagnostic::error(message).with_label(expression.span()));
            }
        }
    }
}

fn resolve_camera_parameter<'a>(
    expression: Option<&'a oxc_ast::ast::Expression<'a>>,
    ctx: &LintContext<'a>,
) -> Option<(&'a oxc_ast::ast::Expression<'a>, f64)> {
    expression.and_then(|expression| {
        resolve_static_number(expression, ctx).map(|value| (expression, value))
    })
}

fn jsx_attribute_expression<'a>(
    attribute: &'a oxc_ast::ast::JSXAttribute<'a>,
) -> Option<&'a oxc_ast::ast::Expression<'a>> {
    let JSXAttributeValue::ExpressionContainer(container) = attribute.value.as_ref()? else {
        return None;
    };
    container.expression.as_expression()
}

fn is_r3f_canvas<'a>(
    opening_element: &oxc_ast::ast::JSXOpeningElement<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    R3F_PUBLIC_MODULES.iter().any(|module_source| {
        resolve_imported_jsx_component_name(opening_element, module_source, ctx) == Some("Canvas")
    })
}
